Doc/reference/expressions.rst

```diff
@@ -183,11 +183,20 @@ by considering each of the :keyword:`for` or :keyword:`if` clauses a block,
 nesting from left to right, and evaluating the expression to produce an element
 each time the innermost block is reached.
 
-Note that the comprehension is executed in a separate scope, so names assigned
-to in the target list don't "leak" into the enclosing scope.
+However, aside from the leftmost :keyword:`for` clause, the comprehension is
+executed in a separate implicitly nested scope. This ensures that names
+assigned to in the target list don't "leak" into the enclosing scope.
 
-Yield expressions are not allowed in comprehensions
-except the expression for the outermost iterable.
+The leftmost :keyword:`for` clause is evaluated directly in the enclosing scope
+and then passed as an argument to the implictly nested scope. Subsequent
+:keyword:`for` clauses cannot be evaluated in the enclosing scope since they
+may depend on the previous :keyword:`for` loop. For example:
+``[x*y for x in range(10) for y in bar(x)]``.
```

Comment:
Maybe use more concrete real-world example? E.g. `for y in range(x+1, 10)`.

Reply:
I lifted this one directly from the generator expression docs, but changing it in both places would definitely be reasonable.